pkg/util/workqueue/delaying_queue_test.go

```diff
@@ -0,0 +1,140 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package workqueue
+
+import (
+	"fmt"
+	"reflect"
+	"testing"
+	"time"
+
+	"k8s.io/kubernetes/pkg/util"
+	"k8s.io/kubernetes/pkg/util/wait"
+)
+
+func TestSimpleQueue(t *testing.T) {
+	fakeClock := util.NewFakeClock(time.Now())
+	q := newDelayingQueue(fakeClock)
+
+	first := "foo"
+
+	q.AddAfter(first, 50*time.Millisecond)
+
+	if q.Len() != 0 {
+		t.Errorf("should not have added")
+	}
+
+	fakeClock.Step(60 * time.Millisecond)
+
+	err := wait.Poll(1*time.Millisecond, 30*time.Millisecond, func() (done bool, err error) {
+		if q.Len() == 1 {
+			return true, nil
+		}
+
+		return false, nil
+	})
+	if err != nil {
+		t.Errorf("should have added")
+	}
+	item, _ := q.Get()
+	q.Done(item)
+
+	// step past the next heartbeat
+	fakeClock.Step(10 * time.Second)
+
+	err = wait.Poll(1*time.Millisecond, 30*time.Millisecond, func() (done bool, err error) {
```

Comment:
If you have a test that is exercising code that is dependent on timing and
threads, it will flake when run in parallel with other tests.  You should
assume you can pause any go routine for multiple seconds and the test
should still pass.

On Wed, Apr 6, 2016 at 7:40 AM, David Eads <notifications@github.com> wrote:

> In pkg/util/workqueue/delaying_queue_test.go
> <https://github.com/kubernetes/kubernetes/pull/23444#discussion_r58690463>
> :
>
> > +		if q.Len() == 1 {
> > +			return true, nil
> > +		}
> > +
> > +		return false, nil
> > +	})
> > +	if err != nil {
> > +		t.Errorf("should have added")
> > +	}
> > +	item, _ := q.Get()
> > +	q.Done(item)
> > +
> > +	// step past the next heartbeat
> > +	fakeClock.Step(10 * time.Second)
> > +
> > +	err = wait.Poll(1*time.Millisecond, 30*time.Millisecond, func() (done bool, err error) {
>
> Same for all of the polls.
>
> Ok, how long would you like? 30 milliseconds is several hundred times
> longer than I've seen.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/23444/files/8f4ffde62153c99cd7faf0f546646363bf10e68d#r58690463>
>